lib/Backend/BackwardPass.cpp

```diff
@@ -6988,19 +6988,31 @@ BackwardPass::ProcessBailOnNoProfile(IR::Instr *instr, BasicBlock *block)
         return true;
     }
 
+    // Save the head instruction for later use.
+    IR::Instr *blockHeadInstr = curInstr;
+
     // We can't bail in the middle of a "tmp = CmEq s1, s2; BrTrue tmp" turned into a "BrEq s1, s2",
     // because the bailout wouldn't be able to restore tmp.
     IR::Instr *curNext = curInstr->GetNextRealInstrOrLabel();
+    IR::Instr *instrNope = nullptr;
     if (curNext->m_opcode == Js::OpCode::Ld_A && curNext->GetDst()->IsRegOpnd() && curNext->GetDst()->AsRegOpnd()->m_fgPeepTmp)
     {
-        block->RemoveInstr(instr);
-        return true;
+        // Instead of just giving up, we can be a little trickier. We can instead treat the tmp declaration(s) as a
+        // part of the block prefix, and put the bailonnoprofile immediately after them. This has the added benefit
+        // that we can still merge up blocks beginning with bailonnoprofile, even if they would otherwise not allow
+        // us to, due to the fact that these tmp declarations would be pre-empted by the higher-level bailout.
+        while (curNext->m_opcode == Js::OpCode::Ld_A && curNext->GetDst()->IsRegOpnd() && curNext->GetDst()->AsRegOpnd()->m_fgPeepTmp)
+        {
+            instrNope = curNext;
+            curNext = curNext->GetNextRealInstrOrLabel();
+        }
+        instrNope = curNext;
     }
 
     curInstr = instr->m_prev;
 
     // Move to top of block.
-    while(!curInstr->StartsBasicBlock())
+    while(!curInstr->StartsBasicBlock() && curInstr != instrNope)
```

Comment:
Tracing the code seems to indicate that it has the desired effect; while they will be equal in cases where the BailOnNoProfile is already just under the Ld_A fgPeepTmps, in many cases there will be other instructions in between. In that case, instrNope serves as a second "start of block", at which the search back towards the start of the block should stop.